Implement getNearest methods

matkt opened the pull request:
## PR description

Add getNearestAfter method needed for verkle and fix getNearestBefore method


## Fixed Issue(s)
<!-- Please link to fixed issue(s) here using format: fixes #<issue number> -->
<!-- Example: "fixes #2" -->


### Thanks for sending a pull request! Have you done the following?

- [ ] Checked out our [contribution guidelines](https://github.com/hyperledger/besu/blob/main/CONTRIBUTING.md)?
- [ ] Considered documentation and added the `doc-change-required` label to this PR [if updates are required](https://wiki.hyperledger.org/display/BESU/Documentation).
- [ ] Considered the changelog and included an [update if required](https://wiki.hyperledger.org/display/BESU/Changelog).
- [ ] For database changes (e.g. KeyValueSegmentIdentifier) considered compatibility and performed forwards and backwards compatibility tests

### Locally, you can run these tests to catch failures early:

- [ ] unit tests: `./gradlew build`
- [ ] acceptance tests: `./gradlew acceptanceTest`
- [ ] integration tests: `./gradlew integrationTest`
- [ ] reference tests: `./gradlew ethereum:referenceTests:referenceTests`

garyschulte commented on services/kvstore/src/main/java/org/hyperledger/besu/services/kvstore/KeyComparator.java:
```diff
@@ -0,0 +1,52 @@
+        return compare;
+      }
+    }
+    return Integer.compare(key1.size(), key2.size());
```
is Byte.compareUnsigned ignoring trailing bytes and/or zero's?

matkt replied:
Not sure to understand why you want to ignore zero in this case

garyschulte replied:
I was curious why the fallback return of Integer.compare - thinking about it today, that is just delegating in case one has a length of 0 right

garyschulte commented on services/kvstore/src/main/java/org/hyperledger/besu/services/kvstore/LayeredKeyValueStorage.java:
```diff
@@ -101,25 +102,72 @@ public Optional<byte[]> get(final SegmentIdentifier segmentId, final byte[] key)
+      if (ourCommonPrefixLength != parentCommonPrefixLength) {
+        return ourCommonPrefixLength > parentCommonPrefixLength ? ourNearest : parentNearest;
+      } else {
+        return (isAfter ^ (ourNearest.get().key().compareTo(parentNearest.get().key()) > 0))
+            ? ourNearest
+            : parentNearest;
```
this is very succinct 🥇, but I had to think hard about this.  It might be worth a comment to explain to our future selves

matkt replied:
Is it just this line that seems complicated or the method?

matkt replied:
I simplified this code and added explanation , feel free to tell me if it's better

garyschulte commented on services/kvstore/src/main/java/org/hyperledger/besu/services/kvstore/SegmentedInMemoryKeyValueStorage.java:
```diff
@@ -143,27 +145,70 @@ public Optional<byte[]> get(final SegmentIdentifier segmentIdentifier, final byt
+          this.hashValueStore.computeIfAbsent(segmentIdentifier, s -> newSegmentMap());
+
+      // Define a function to find the nearest key-value pair based on a given predicate
+      final Function<Predicate<Map.Entry<Bytes, Optional<byte[]>>>, Optional<NearestKeyValue>>
+          findNearest =
+              (predicate) -> {
+                // Filter the segment map's entries based on the predicate
+                final Stream<Map.Entry<Bytes, Optional<byte[]>>> filteredStream =
+                    segmentMap.entrySet().stream().filter(predicate);
+                // Depending on the useMin flag, find either the minimum or maximum entry according
+                // to key order
+                final Optional<Map.Entry<Bytes, Optional<byte[]>>> sortedStream =
+                    useMin
+                        ? filteredStream.min(Map.Entry.comparingByKey())
+                        : filteredStream.max(Map.Entry.comparingByKey());
+                // Map the found entry to a NearestKeyValue object
+                return sortedStream.map(
+                    entry -> new NearestKeyValue(entry.getKey(), entry.getValue()));
+              };
```
The only param from the method signature used in this function is useMin.  Can we define 2 static predicates, one for before and one for after, rather than creating a new function for every call to getNearest?    I would presume that allocating a new function for every call to getNearest is unnecessarily expensive

matkt replied:
we are also using segmentMap and this one is dynamic . not sure how you want to do, if you can give me an example to understand your idea

garyschulte replied:
it might or might not be more expensive  to create a function for every call versus just having a static function we pass parameters to.  It would be an interesting question to answer with profiling

matkt replied:
maybe @ahamlat can help for that ? if you think it will impact performance I will modify this part

ahamlat replied:
We can do micro-benchmarking with JMH on both implementations and enable Async profiler during the test. I just need to have an example of the usage of both methods, like a unit test.

garyschulte commented on services/kvstore/src/test/java/org/hyperledger/besu/services/kvstore/AbstractSegmentedKeyValueStorageTest.java:
```diff
@@ -29,52 +30,58 @@
           .forEach(
               i -> {
                 final byte[] key = bytesFromHexString("000" + i);
+                System.out.println(Bytes.of(key));
```
☠️

garyschulte commented on services/kvstore/src/test/java/org/hyperledger/besu/services/kvstore/AbstractSegmentedKeyValueStorageTest.java:
```diff
@@ -29,52 +30,58 @@
                 tx.put(SEGMENT_IDENTIFIER, key, value);
                 // different common prefix, and reversed order of bytes:
                 final byte[] key2 = bytesFromHexString("010" + (10 - i));
+                System.out.println(Bytes.of(key2));
```
☠️

matkt replied:
removed

garyschulte reviewed (changes requested):
Generally LGTM.  the rewrite of getNearest seems a little chat-gpt-y to me though

matkt commented:
> Generally LGTM. the rewrite of getNearest seems a little chat-gpt-y to me though

Thanks for the review.. Can you also tell me what bothers so that I can change. Is factorization too hard to understand ?

The use of Min Max on this case is more interesting because it avoids sorting that would be more expensive. This is why this proposal seems to me more interesting for this call. But I can understand that the overall factorization of the method is complex and I can try to simplify by duplicating more

matkt commented:
@garyschulte did some clean in the PR, I also modified the tests in order to check more edgcases, the new tests are comparing directly with rocksdb instead of hardcoding the responses

thanks to that I fixed some new issues

garyschulte commented:
> The use of Min Max on this case is more interesting because it avoids sorting that would be more expensive.

that makes a lot of sense, min/max will be O(n). 👍

garyschulte reviewed (approved):
LGTM.  I am interested to know if there is measurable overhead for creating a function on every call to getNearest, but it is not worth blocking the PR for.

macfarla commented:
@garyschulte is this GTG?

matkt commented:
yes will fix the PR
